homeassistant/components/diagnostics/__init__.py

```diff
@@ -0,0 +1,107 @@
+"""The Diagnostics integration."""
+from __future__ import annotations
+
+import json
+import logging
+
+from aiohttp import web
+import voluptuous as vol
+
+from homeassistant.components import http, websocket_api
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import integration_platform
+from homeassistant.helpers.json import ExtendedJSONEncoder
+from homeassistant.util.json import (
+    find_paths_unserializable_data,
+    format_unserializable_data,
+)
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up Diagnostics from a config entry."""
+    hass.data[DOMAIN] = {}
+
+    await integration_platform.async_process_integration_platforms(
+        hass, DOMAIN, _register_diagnostics_platform
+    )
+
+    websocket_api.async_register_command(hass, handle_info)
+    hass.http.register_view(DownloadDiagnosticsView)
+
+    return True
+
+
+async def _register_diagnostics_platform(hass, integration_domain, platform):
+    """Register a diagnostics platform."""
+    hass.data[DOMAIN][integration_domain] = {
+        "config_entry": getattr(platform, "async_get_config_entry_diagnostics", None)
```

Comment:
We could define a protocol type for the expected platform module.

Maybe we should blow up if the expected function is missing? Ie, not use `getattr`.

Reply:
Each handler will be optional. In the future we can have device or integration too.